### ISSUE TITLE
Assigning a value to COLOR in a spatial fragment shader causes freeze & console error spam

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#13022]
Godot 3.0 master
Windows 10 64 bits
nVidia 940M

I tried writing this in a spatial shader, which might be wrong:

```glsl
void fragment() {
	COLOR = texture(color_texture, UV);
}
```

However, instead of highlighting the line in red, Godot froze for about 10 seconds, my mesh disappeared and I had to look at the system console, scroll up a huge spam of errors to see what was going on, and guess that assigning COLOR is bad:

```
ERROR: _display_error_with_code: SceneShaderGLES3: Fragment Program Compilation Failed:
0(1164) : error C7565: assignment to varying 'color_interp'

   At: drivers\gles3\shader_gles3.cpp:167
ERROR: ShaderGLES3::get_current_version: SceneShaderGLES3: Fragment Program Compilation Failed:
0(1164) : error C7565: assignment to varying 'color_interp'

   At: drivers\gles3\shader_gles3.cpp:470
ERROR: ShaderGLES3::get_current_version: Method/Function Failed, returning: 0
   At: drivers\gles3\shader_gles3.cpp:478
ERROR: ShaderGLES3::bind: Condition ' !version ' is true. returned: false
   At: drivers\gles3\shader_gles3.cpp:122
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: D:\PROJETS\INFO\GODOT\ENGINE\godot_fork\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: D:\PROJETS\INFO\GODOT\ENGINE\godot_fork\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: D:\PROJETS\INFO\GODOT\ENGINE\godot_fork\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: D:\PROJETS\INFO\GODOT\ENGINE\godot_fork\drivers/gles3/shader_gles3.h:376
ERROR: ShaderGLES3::_get_uniform: Condition ' !version ' is true. returned: -1
   At: D:\PROJETS\INFO\GODOT\ENGINE\godot_fork\drivers/gles3/shader_gles3.h:376
...
...
```

It repeats all this everytime the editor redraws (my scene has a consequent amount of mesh instances using the same shader).
There should be a better way to tell a user that the shader is invalid.

**Comment by LinuxUserGD** [comment]
@Zylann I can partly reproduce this issue
`shader_type spatial;

void fragment() {
	**uniform sampler2D grass;**
	uniform sampler2D rock;
	uniform sampler2D dirt;
	uniform sampler2D splatmap;
	
	vec3 result;
	vec3 grasscol;
	vec3 rockcol;
	vec3 dirtcol;
	
	float grassval = texture(splatmap,UV).g;
	float rockval = texture(splatmap,UV).r;
	float dirtval = texture(splatmap,UV).b;
	
	grasscol = texture(grass,UV).rgb * grassval;
	rockcol = texture(rock,UV).rgb * rockval;
	dirtcol = texture(dirt,UV).rgb * dirtval;
	
	result = grasscol+rockcol+dirtcol;
}
`
Output:
` :4 - Expected expression, found: OUT
 drivers/gles3/rasterizer_storage_gles3.cpp:1671 - Condition ' err != OK ' is true.
`
First line is marked
I get the same error for every hint (vec4, int, float, sampler2D)

When I remove "uniform":

`shader_type spatial;

void fragment() {
	sampler2D grass;
	sampler2D rock;
	sampler2D dirt;
	sampler2D splatmap;
	
	vec3 result;
	vec3 grasscol;
	vec3 rockcol;
	vec3 dirtcol;
	
	float grassval = texture(splatmap,UV).g;
	float rockval = texture(splatmap,UV).r;
	float dirtval = texture(splatmap,UV).b;
	
	grasscol = texture(grass,UV).rgb * grassval;
	rockcol = texture(rock,UV).rgb * rockval;
	dirtcol = texture(dirt,UV).rgb * dirtval;
	
	result = grasscol+rockcol+dirtcol;
}
`
Output (only errors because it's over 17k lines long:

` drivers/gles3/shader_gles3.cpp:167 - SceneShaderGLES3: Fragment Program Compilation Failed:
0:1163(2): error: opaque variables must be declared uniform
0:1164(2): error: opaque variables must be declared uniform
0:1165(2): error: opaque variables must be declared uniform
0:1166(2): error: opaque variables must be declared uniform
0:1171(27): warning: `m_splatmap' used uninitialized
0:1172(26): warning: `m_splatmap' used uninitialized
0:1173(26): warning: `m_splatmap' used uninitialized
0:1174(22): warning: `m_grass' used uninitialized
0:1175(21): warning: `m_rock' used uninitialized
0:1176(21): warning: `m_dirt' used uninitialized

 drivers/gles3/shader_gles3.cpp:440 - SceneShaderGLES3: Fragment Program Compilation Failed:
0:1163(2): error: opaque variables must be declared uniform
0:1164(2): error: opaque variables must be declared uniform
0:1165(2): error: opaque variables must be declared uniform
0:1166(2): error: opaque variables must be declared uniform
0:1171(27): warning: `m_splatmap' used uninitialized
0:1172(26): warning: `m_splatmap' used uninitialized
0:1173(26): warning: `m_splatmap' used uninitialized
0:1174(22): warning: `m_grass' used uninitialized
0:1175(21): warning: `m_rock' used uninitialized
0:1176(21): warning: `m_dirt' used uninitialized

 drivers/gles3/shader_gles3.cpp:448 - Method/Function Failed, returning: __null
 drivers/gles3/shader_gles3.cpp:122 - Condition ' !version ' is true. returned: false
 ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1
 ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1
 ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1
 ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1
 ./drivers/gles3/shader_gles3.h:376 - Condition ' !version ' is true. returned: -1
`
Is this normal / my shader code wrong or should i report a new issue?

**Comment by LinuxUserGD** [comment]
Both outputs contain the "Program Compilation Failes" error, but it didn't freeze on Linux

**Comment by LinuxUserGD** [comment]
[splatmap.zip](https://github.com/godotengine/godot/files/1592073/splatmap.zip)

**Comment by LinuxUserGD** [comment]
@Noshyaar  Could you add a bug label please? Should I open a new issue?

**Comment by binbitten** [comment]
Got it. In a spatial shader "COLOR" is translated to "color_interp" in glsl. This is a varying and cannot be assigned in fragments. This was fixed in https://github.com/godotengine/godot/commit/e946fb8cb0a9177c45db5305a26efa52ea8e7907 (at least partially, members of a constant has a fix in  #15219).

LinuxUserGD's issue is unrelated.

**Comment by akien-mga** [comment]
So fixed by e946fb8c IIUC.

@LinuxUserGD Please open another issue if you can still reproduce your bug in the current master branch.
